apps/website/src/server/utils/encryption.ts

```
@@ -0,0 +1,124 @@
+    "raw",
+    encoder.encode(key),
+    { name: KEY_ALGORITHM_NAME },
+    false,
+    ["deriveKey"]
+  );
+}
+
+export async function deriveSaltedEncryptionKey(
+  key: CryptoKey,
+  salt: ArrayBuffer
+) {
+  return crypto.subtle.deriveKey(
+    {
+      name: KEY_ALGORITHM_NAME,
+      salt,
+      iterations: HASH_ITERATIONS,
+      hash: HASH_ALGORITHM_NAME,
+    },
+    key,
+    { name: ENCRYPTION_ALGORITHM_NAME, length: ENCRYPTION_ALGORITHM_LENGTH },
+    false,
+    ["encrypt", "decrypt"]
+  );
+}
+
+export async function createSaltedEncryptionKey(salt: ArrayBuffer) {
+  const dataEncryptionKey = await importEncryptionKey(
+    env.DATA_ENCRYPTION_PASSPHRASE
+  );
+  return deriveSaltedEncryptionKey(dataEncryptionKey, salt);
+}
+
+export async function encrypt(
+  plainText: string,
+  saltedEncryptionKey: CryptoKey
+) {
+  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH));
+  const plainTextBuffer = encoder.encode(plainText);
+
+  const encryptedCipherTextBuffer = await crypto.subtle.encrypt(
+    { name: ENCRYPTION_ALGORITHM_NAME, iv, tagLength: 128 },
+    saltedEncryptionKey,
+    plainTextBuffer
+  );
+
+  const resultBuffer = new Uint8Array(
+    iv.length + encryptedCipherTextBuffer.byteLength
+  );
+  resultBuffer.set(iv);
+  resultBuffer.set(new Uint8Array(encryptedCipherTextBuffer), iv.length);
+
+  return base64Encode(resultBuffer);
+}
+
+export async function decrypt(
+  encryptedText: string,
+  saltedEncryptionKey: CryptoKey
+) {
+  if (encryptedText === "") {
+    return "";
+  }
+
+  try {
+    const encryptedBuffer = base64Decode(encryptedText);
+
+    const iv = encryptedBuffer.slice(0, IV_LENGTH);
+    const cipherText = encryptedBuffer.slice(IV_LENGTH);
+
+    const decryptedBuffer = await crypto.subtle.decrypt(
+      { name: ENCRYPTION_ALGORITHM_NAME, iv, tagLength: 128 },
+      saltedEncryptionKey,
+      cipherText
+    );
+    return decoder.decode(decryptedBuffer);
+  } catch (e) {
+    return "";
+  }
```

Do we not want to surface an error if a bad decryption happens? I worry this risks some level of data loss if a key were accidentally changed etc.?

---

> What's the plan for encrypting existing stuff?
For the giveaway entries i wanted to just let the decryption util allow empty values (+ empty salt), and because we removed all the fields that would be encrypted (= PII) that would be the case for all of them.

> Are we going to do encryption for user-account-related tables too (maybe a separate PR after)?
Sure. Right now that would only be the email address though. Not sure if we need encryption for that at the cost of making auth potentially slower.